File: golang/78.go
package golang
func subsets(nums []int) [][]int {
    result := &[][]int{}
    dps(nums, result, []int{}, 0)
    return *result
}

func dps (nums []int, result *[][]int, current []int, start int) {
    *result = append(*result, append([]int{}, current...))
    for i:= start; i<len(nums); i++ {
        current = append(current, nums[i])
        dps(nums, result, current, i + 1)
        current = current[:len(current) - 1]
    }
}

File: java/78.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        dfs(nums, result, new ArrayList<>(), 0);
        return result; 
    }

    private void dfs(int[] nums, List<List<Integer>> result, List<Integer> currentList, int start) {
        result.add(new ArrayList<>(currentList));
        for (int i=start; i<nums.length; i++) {
            currentList.add(nums[i]);
            dfs(nums, result, currentList, i+1);
            currentList.remove(currentList.size() - 1);
        }
    }
}